src/iotjs_module_httpparser.cpp

```diff
@@ -0,0 +1,483 @@
+    reinterpret_cast<HTTPParserWrap*>(jparser->GetNative());
+  parser->Initialize(httpparser_type);
+
+  return true;
+}
+
+
+JHANDLER_FUNCTION(Finish, handler) {
+  IOTJS_ASSERT(handler.GetArgLength() == 0);
+  IOTJS_ASSERT(handler.GetThis()->IsObject());
+  JObject* jparser = handler.GetThis();
+  HTTPParserWrap* parser =
+    reinterpret_cast<HTTPParserWrap*>(jparser->GetNative());
+
+  int rv = http_parser_execute(&(parser->parser), &settings, NULL, 0);
+
+  if (rv != 0) {
+    enum http_errno err = HTTP_PARSER_ERRNO(&parser->parser);
+
+    JObject eobj(JObject::Error("Parse Error"));
+    JSETPROPERTY(eobj, "byteParsed", 0);
+    JSETPROPERTY(eobj, "code", http_errno_name(err));
+    handler.Return(eobj);
+  }
+
+  return true;
+}
+
+
+JHANDLER_FUNCTION(Execute, handler) {
+  IOTJS_ASSERT(handler.GetArgLength() == 1);
+  IOTJS_ASSERT(handler.GetThis()->IsObject());
+  IOTJS_ASSERT(handler.GetArg(0)->IsObject());
+
+  JObject* jparser = handler.GetThis();
+  HTTPParserWrap* parser =
+    reinterpret_cast<HTTPParserWrap*>(jparser->GetNative());
+
+
+  JObject* jbuffer = handler.GetArg(0);
+  BufferWrap* buffer = BufferWrap::FromJBuffer(*jbuffer);
+  char* buf_data = buffer->buffer();
+  int buf_len = buffer->length();
+
+
+  parser->cur_jbuf = jbuffer;
+  parser->cur_buf = buf_data;
+  parser->cur_buf_len = buf_len;
+  parser->had_exception = false;
+
+  int nparsed = http_parser_execute(&(parser->parser), &settings,
+                                    buf_data, buf_len);
+
+  parser->cur_jbuf = NULL;
+  parser->cur_buf = NULL;
+  parser->cur_buf_len = 0;
+
+
+  if (parser->had_exception) {
+    return false;
+  }
+
+  if (nparsed != buf_len) {
+    enum http_errno err = HTTP_PARSER_ERRNO(&parser->parser);
+    JObject eobj(JObject::Error("Parse Error"));
+    JSETPROPERTY(eobj, "byteParsed", 0);
+    JSETPROPERTY(eobj, "code", http_errno_name(err));
+    handler.Return(eobj);
+  }
+  else{
+    JObject ret(nparsed);
+    handler.Return(ret);
+  }
+
+  return true;
+}
+
+
+JHANDLER_FUNCTION(Pause, handler) {
+  IOTJS_ASSERT(handler.GetArgLength() == 0);
+  IOTJS_ASSERT(handler.GetThis()->IsObject());
+  JObject* jparser = handler.GetThis();
+  HTTPParserWrap* parser =
+    reinterpret_cast<HTTPParserWrap*>(jparser->GetNative());
+  http_parser_pause(&parser->parser, 1);
+  return true;
+}
+
+
+JHANDLER_FUNCTION(Resume, handler) {
+  IOTJS_ASSERT(handler.GetArgLength() == 0);
+  IOTJS_ASSERT(handler.GetThis()->IsObject());
+  JObject* jparser = handler.GetThis();
+  HTTPParserWrap* parser =
+    reinterpret_cast<HTTPParserWrap*>(jparser->GetNative());
+  http_parser_pause(&parser->parser, 0);
+  return true;
+}
+
+
+JHANDLER_FUNCTION(Close, handler) {
+  IOTJS_ASSERT(handler.GetArgLength() == 0);
+  IOTJS_ASSERT(handler.GetThis()->IsObject());
+
+  JObject* jparser = handler.GetThis();
+  HTTPParserWrap* parser =
+    reinterpret_cast<HTTPParserWrap*>(jparser->GetNative());
+
+  delete parser;  // FIXME: used in FreeParser.
```

Comment:
I don't think deleting parser instance here is safe. This may lead double free.
Fortunately it looks like there is no place where `parser.close()` is called.